File: ruby/lib/runapi/wan/resources/speech_to_video.rb
# frozen_string_literal: true

module RunApi
  module Wan
    module Resources
      # Generates lip-synced talking-head videos from a portrait image and
      # driving speech audio. Both source_image_url and source_audio_url are required.
      class SpeechToVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/wan/speech_to_video"
        RESPONSE_CLASS = Types::VideoTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedVideoTaskResponse

        def initialize(http)
          @http = http
        end

        # Generate a lip-synced video and wait until complete.
        #
        # @param params [Hash] speech-to-video parameters
        # @return [RunApi::Wan::Types::CompletedVideoTaskResponse] completed video generation
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create a speech-to-video generation task.
        #
        # @param params [Hash] speech-to-video parameters
        # @return [RunApi::Wan::Types::VideoTaskResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get speech-to-video status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Wan::Types::VideoTaskResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["speech-to-video"], params)
        end
      end
    end
  end
end

File: python/src/runapi/wan/resources/speech_to_video.py
"""Wan speech-to-video resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedVideoTaskResponse,
    VideoTaskResponse,
)


class SpeechToVideo(Resource):
    """Generate speech-driven talking-head videos with Wan models."""

    ENDPOINT = "/api/v1/wan/speech_to_video"

    RESPONSE_CLASS = VideoTaskResponse
    COMPLETED_RESPONSE_CLASS = CompletedVideoTaskResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Generate a lip-synced video and poll until it completes.

        Args:
            **params: speech-to-video parameters (model, ...).

        Returns:
            The completed (narrowed) speech-to-video response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a speech-to-video task and return immediately with an id.

        Args:
            **params: speech-to-video parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["speech-to-video"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a speech-to-video task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
